clustermesh-apiserver/clustermesh/synchronizer.go

```diff
@@ -66,45 +84,128 @@ func RegisterSynchronizer[T runtime.Object](in syncParams[T]) {
 	store := in.Factory.NewSyncStore(
 		in.ClusterInfo.Name, in.Client,
 		in.Options.Prefix, in.Options.StoreOpts...)
+
 	synced := in.SyncState.WaitForResource()
 
+	// process is a helper function to log and execute store operations.
+	process := func(invoker, op, key string, do func() error) {
+		logger.Info("Updating resource in etcd",
+			logfields.LogSubsys, invoker,
+			logfields.Operation, op,
+			logfields.Key, key,
+		)
+		if err := do(); err != nil {
+			logger.Warn("Failed updating resource in etcd",
+				logfields.Error, err,
+				logfields.LogSubsys, invoker,
+				logfields.Operation, op,
+				logfields.Key, key,
+			)
+		}
+	}
+
 	in.JobGroup.Add(
 		job.OneShot(
 			fmt.Sprintf("%s-sync", strings.ToLower(in.Options.Resource)),
 			func(ctx context.Context, _ cell.Health) error {
-				for event := range in.Resource.Events(ctx) {
-					event.Done(nil)
+				resourceStore, err := in.Resource.Store(ctx)
+				if err != nil {
+					return err
+				}
+				invoker := fmt.Sprintf("%s-sync", strings.ToLower(in.Options.Resource))
+				logger.Info("Starting event synchronizer")
 
-					if event.Kind == resource.Sync {
-						logger.Info("Initial entries successfully received from Kubernetes")
-						store.Synced(ctx, synced)
-						continue
-					}
+				// Get event channels
+				resourceEvents := in.Resource.Events(ctx)
+				var namespaceEvents <-chan resource.Event[*slim_corev1.Namespace]
+				if in.Options.Namespaced {
+					logger.Debug("Namespace watcher is enabled for resource type")
+					namespaceEvents = in.Namespaces.Events(ctx)
+				} else {
+					logger.Debug("Namespace watcher is not enabled for resource type")
+				}
 
-					process := func(op, key string, do func() error) {
-						logger.Info("Updating resource in etcd",
-							logfields.Operation, op,
-							logfields.Key, key,
-						)
-
-						if err := do(); err != nil {
-							logger.Warn("Failed updating resource in etcd",
-								logfields.Error, err,
-								logfields.Operation, op,
-								logfields.Key, key,
-							)
+				for {
+					select {
+					case <-ctx.Done():
+						logger.Info("Context done, stopping event synchronizer")
+						return nil
```

Comment:
Drop the `ctx.Done` case, as the other channels are already closed when the context is canceled. That ensures that `event.Done` is called on all events (otherwise the associated finalizer would complain, possibly causing test failures).

Reply:
To clarify - we would wait for main thread to kill this job instead of exiting ourselves?

Reply:
Ok, that makes sense. I have added a `default` case

```
...
default:
    if resourceEvents == nil && namespaceEvents == nil {
	    logger.Info("All event channels closed, stopping synchronizer")
	    return nil
    }
...
```